README.md
SMS and Call Filter
==================


SMS and Call filter is a fork project of SMS Filter by Jelly Greets.


License
=======


>  Author: Hasib Al Muhaimin.

>  The contents of this file are subject to the Mozilla Public License
  Version 1.1 (the "License"); you may not use this file except in
  compliance with the License. You may obtain a copy of the License at
  http://www.mozilla.org/MPL/

  ***Courtesy to Jelle Geerts:***
>  Author: Jelle Geerts

>  Usage of the works is permitted provided that this instrument is
  retained with the works, so that any entity that uses the works is
  notified of this instrument.

>  DISCLAIMER: THE WORKS ARE WITHOUT WARRANTY.


Features:
=========


 * Just Install and enjoy.
 * If Mobile no is less than 5 digits, It will get blocked.
 * It will block a SMS if it think the sms is spam/promotional/unusual.
 * User can turn off notification.
 * Do block sms from user who hide their mobile no.
 * User can add Custom Filters for messages.
 * Can restore messages to inbox.
 * No ad.

Change Log:
==========

####Version 1.4.8
* Few Numbers has been added in the block list by default. For incoming call only.

####Version 1.4.7
* Update Weight vector.
* Change of Hash Function and HashFullString Function.
  Idea of Bloom Filter added.
* Message can be restored from MessageViewer.
* Major change in SKSpamBlocker.java.


####Version 1.4.6
* Update Weight vector for Neural Netowork. (reduced overlapping
  messages in DATABASE for training)
* Show Notification on Delete message from 'MessageList' and
  show notification on any error while restoring message to
  inbox.


####Version 1.4.5
* Update Weight vector for Neural Netowork.

####Version 1.4.4

* Now call will get blocked if number is less than '5' digits. Previously the threshold was *less than '6' digits*.
* Change License.
* Minor Bug fix.

####Version 1.4.3

* Blocks sms from user who hides their mobile no.


####Version 1.4.2

* Added about page with Licence and courtesy.



####Version 1.4.1

* Fix a bug in pendingintent. now the request id will be equal to
  message id mod 2^28.

####Version 1.4


* Fix a bug about PendingIntent for Notification bar.
* Added Neural Network.
* Added feature to restore message to inbox.
* Change of app Icon.
* Change feature. A spam message will always be saved
  User can only change settings to show notification or not.
* Added call filter. New permission is needed from now on.

####Version 1.3


* Fix code so it still compiles on Android 2.2.

####Version 1.2


* Show a warning when the application is run on Android 4.4 or newer,
  as the application doesn't work on 4.4, and possibly doesn't work on
  newer versions either.
  See https://code.google.com/p/android/issues/detail?id=61684 .

####Version 1.1


* New feature: entering #ANY# as the filter address creates a filter
  that applies to any address.

####Version 1.0


* First release.


ChangeLog.txt
Version 1.4.7 
=========================================
* Update Weight vector.
* Change of Hash Function and HashFullString Function. 
  Idea of Bloom Filter added.
* Message can be restored from MessageViewer.
* Major change in SKSpamBlocker.java.


Version 1.4.6 (October 10, 2014. 07:19 PM):
==========================================
* Update Weight vector for Neural Netowork. (reduced overlapping
  messages in DATABASE for training)
* Show Notification on Delete message from 'MessageList' and
  show notification on any error while restoring message to
  inbox.

Version 1.4.5 (October 9, 2014. 08:19 PM):
==========================================
* Update Weight vector for Neural Netowork.

Version 1.4.4 (October 3, 2014. 05:59PM):
=========================================
* Now call will get blocked if number is less than '5' digits. Previously the threshold was *less than '6' digits*.
* Change License.
* Minor Bug fix.

Version 1.4.3 (October 2, 2014. 12:18AM):
=========================================
* Blocks sms from user who hides their mobile no.


Version 1.4.2 (October 1, 2014. 4:18 AM):
=========================================
* Added about page with Licence and courtesy.



Version 1.4.1 (September 29, 2014. 5:00 PM):
============================================
* Fix a bug in pendingintent. now the request id will be equal to 
  message id mod 2^28.

Version 1.4 (September 28, 2014. 2:56 AM): 
==========================================

* Fix a bug about PendingIntent for Notification bar.
* Added Neural Network.
* Added feature to restore message to inbox.
* Change of app Icon.
* Change feature. A spam message will always be saved
  User can only change settings to show notification or not.
* Added call filter. New permission is needed from now on.

Version 1.3 (August 26, 2014):
==============================

* Fix code so it still compiles on Android 2.2.

Version 1.2 (August 25, 2014):
==============================

* Show a warning when the application is run on Android 4.4 or newer,
  as the application doesn't work on 4.4, and possibly doesn't work on
  newer versions either.
  See https://code.google.com/p/android/issues/detail?id=61684 .

Version 1.1 (November 22, 2013):
================================

* New feature: entering #ANY# as the filter address creates a filter
  that applies to any address.

Version 1.0 (January 22, 2013):
===============================

* First release.


src/com/sadakhata/smsandcallfilter/Settings.java
/*
 * Author: Hasib Al Muhaimin.
 * 
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 * 
 * ***courtesy to Jelle Geerts***
 * Author: Jelle Geerts

 * Usage of the works is permitted provided that this instrument is
 * retained with the works, so that any entity that uses the works is
 * notified of this instrument.
 *
 * DISCLAIMER: THE WORKS ARE WITHOUT WARRANTY.
*/

package com.sadakhata.smsandcallfilter;

import java.util.ArrayList;
import java.util.List;

import com.sadakhata.smsandcallfilter.R;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;
import android.widget.Toast;

public class Settings extends SQLiteOpenHelper
{
    private static final String TAG = "Settings";

    public static final String ACTION_NEW_MESSAGE = C.PACKAGE_NAME + ".new_message";

    public static final String ANY_ADDRESS = "#ANY#";

    private static final int DATABASE_VERSION = 1;

    private static final String SETTING_SAVE_MESSAGES   = "save_messages";

    private static final String DATABASE_NAME           = C.PACKAGE_NAME + ".db";

    private static final String SETTINGS_TABLE          = "settings";
    private static final String KEY_ID                  = "id";
    private static final String KEY_KEY                 = "key";
    private static final String KEY_VALUE               = "value";

    private static final String FILTERS_TABLE           = "filters";
    private static final String KEY_NAME                = "name";
    private static final String KEY_ADDRESS             = "address";

    private static final String CONTENT_STRINGS_TABLE   = "content_strings";
    private static final String KEY_FILTER_ID           = "filter_" + KEY_ID;

    private static final String MESSAGES_TABLE          = "messages";
    private static final String KEY_RECEIVED_AT         = "received_at";

    private static final String FILTERS_ORDER_BY        = KEY_NAME + " ASC";
    private static final String MESSAGES_ORDER_BY       = KEY_RECEIVED_AT + " DESC";

    private Context mContext;

    public Settings(Context context)
    {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        mContext = context;
    }

    @Override
    public void onCreate(SQLiteDatabase db)
    {
        String sql =
            "CREATE TABLE " + SETTINGS_TABLE + "("
            + KEY_ID + " INTEGER PRIMARY KEY AUTOINCREMENT"
            + ", " + KEY_KEY + " TEXT NOT NULL UNIQUE"
            + ", " + KEY_VALUE + " TEXT)";
        Log.d(TAG, "SQL:\n" + sql);
        db.execSQL(sql);

        sql =
            "CREATE TABLE " + FILTERS_TABLE + "("
            + KEY_ID + " INTEGER PRIMARY KEY AUTOINCREMENT"
            + ", " + KEY_NAME + " TEXT NOT NULL UNIQUE"
            + ", " + KEY_ADDRESS + " TEXT NOT NULL)";
        Log.d(TAG, "SQL:\n" + sql);
        db.execSQL(sql);

        sql =
            "CREATE TABLE " + CONTENT_STRINGS_TABLE + "("
            + KEY_ID + " INTEGER PRIMARY KEY AUTOINCREMENT"
            + ", " + KEY_FILTER_ID + " INTEGER REFERENCES " + FILTERS_TABLE + "(" + KEY_ID + ") ON DELETE CASCADE"
            + ", " + KEY_VALUE + " TEXT NOT NULL)";
        Log.d(TAG, "SQL:\n" + sql);
        db.execSQL(sql);

        sql =
            "CREATE TABLE " + MESSAGES_TABLE + "("
            + KEY_ID + " INTEGER PRIMARY KEY AUTOINCREMENT"
                    + ", " + KEY_ADDRESS + " TEXT NOT NULL"
            + ", " + KEY_RECEIVED_AT + " INTEGER NOT NULL"
            + ", " + KEY_VALUE + " TEXT)";
        Log.d(TAG, "SQL:\n" + sql);
        db.execSQL(sql);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion)
    {
    }

    @Override
    public SQLiteDatabase getWritableDatabase()
    {
        SQLiteDatabase db = super.getWritableDatabase();
        if (!db.isReadOnly())
            db.execSQL("PRAGMA foreign_keys = ON");
        return db;
    }

    public String getSetting(String key, String defaultValue)
    {
        String value = null;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.query(
            SETTINGS_TABLE,
            new String[] { KEY_VALUE },
            KEY_KEY + "=?",
            new String[] { key },
            null, null, null);
        while (cursor.moveToNext())
        {
            if (value != null)
            {
                // There must be only one setting with this key.
                throw new AssertionError();
            }
            value = cursor.getString(0);
        }
        cursor.close();
        db.close();
        if (value == null)
            value = defaultValue;
        return value;
    }

    public void setSetting(String key, String value)
    {
        if (key.length() == 0)
            throw new AssertionError();
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(KEY_KEY, key);
        values.put(KEY_VALUE, value);
        long rowID = db.replaceOrThrow(SETTINGS_TABLE, null, values);
        db.close();
        if (rowID < 0)
            throw new AssertionError();
    }

    public boolean saveMessages()
    {
        final boolean defaultValue = true;
        return Boolean.parseBoolean(getSetting(SETTING_SAVE_MESSAGES, String.valueOf(defaultValue)));
    }

    public void setSaveMessages(boolean value)
    {
        setSetting(SETTING_SAVE_MESSAGES, Boolean.toString(value));
    }

    public boolean isFilterNameUsed(String name)
    {
        return findFilterByName(name) != null;
    }

    public Filter findFilterByName(String name)
    {
        Filter filter = null;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.query(
            FILTERS_TABLE,
            new String[] { KEY_NAME, KEY_ADDRESS },
            KEY_NAME + "=?",
            new String[] { name },
            null, null, null);
        while (cursor.moveToNext())
        {
            if (filter != null)
            {
                // There must be only one filter with this name.
                throw new AssertionError();
            }
            String address = cursor.getString(1);
            List<String> contentFilters = getContentFilters(name);
            filter = new Filter(name, address, contentFilters);
        }
        cursor.close();
        db.close();
        return filter;
    }

    // Find filters that match the given address.
    // The search will _not_ return filters that match _any_ address.
    public List<Filter> findFiltersByAddress(String address)
    {
        List<Filter> filters = new ArrayList<Filter>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.query(
            FILTERS_TABLE,
            new String[] { KEY_NAME, KEY_ADDRESS },
            KEY_ADDRESS + "=?",
            new String[] { address },
            null, null,
            FILTERS_ORDER_BY);
        while (cursor.moveToNext())
        {
            String name = cursor.getString(0);
            List<String> contentFilters = getContentFilters(name);
            filters.add(new Filter(name, address, contentFilters));
        }
        cursor.close();
        db.close();
        return filters;
    }

    // Find filters that match the given address.
    // The search will _also_ return filters that match _any_ address.
    public List<Filter> findFiltersForAddress(String address)
    {
        List<Filter> list = findFiltersByAddress(address);
        list.addAll(getWildcardFilters());
        return list;
    }

    public Filter getFilterByName(String name)
    {
        Filter filter = findFilterByName(name);
        if (filter == null)
            throw new AssertionError();
        return filter;
    }

    // Get filters that are meant to match _any_ address.
    public List<Filter> getWildcardFilters()
    {
        return findFiltersByAddress(ANY_ADDRESS);
    }

    public List<Filter> getFilters()
    {
        List<Filter> filters = new ArrayList<Filter>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.query(
            FILTERS_TABLE,
            new String[] { KEY_NAME, KEY_ADDRESS },
            null, null, null, null,
            FILTERS_ORDER_BY);
        while (cursor.moveToNext())
        {
            String name = cursor.getString(0);
            String address = cursor.getString(1);
            List<String> contentFilters = getContentFilters(name);
            filters.add(new Filter(name, address, contentFilters));
        }
        cursor.close();
        db.close();
        return filters;
    }

    public List<String> getContentFilters(String filterName)
    {
        List<String> contentFilters = new ArrayList<String>();
        SQLiteDatabase db = getReadableDatabase();
        String sql =
            "SELECT s." + KEY_VALUE + " FROM " + FILTERS_TABLE + " f"
            + " JOIN " + CONTENT_STRINGS_TABLE + " s"
            + "  ON f." + KEY_ID + "=s." + KEY_FILTER_ID
            + " WHERE f." + KEY_NAME + "=?";
        Cursor cursor = db.rawQuery(sql, new String[] { filterName });
        while (cursor.moveToNext())
        {
            String value = cursor.getString(0);
            contentFilters.add(value);
        }
        cursor.close();
        db.close();
        return contentFilters;
    }

    public String getFilterAddress(String name)
    {
        Filter filter = getFilterByName(name);
        return filter.address;
    }

    public void saveFilter(Filter filter)
    {
        SQLiteDatabase db = getWritableDatabase();
        try
        {
            if (filter.name.length() == 0)
                throw new AssertionError();
            if (filter.address.length() == 0)
                throw new AssertionError();

            db.beginTransaction();

            ContentValues values = new ContentValues();
            values.put(KEY_NAME, filter.name);
            values.put(KEY_ADDRESS, filter.address);
            long filter_id = db.replaceOrThrow(FILTERS_TABLE, null, values);
            if (filter_id < 0)
                throw new AssertionError();

            for (String s : filter.contentFilters)
            {
                if (s.length() == 0)
                    throw new AssertionError();
                values.clear();
                values.put(KEY_FILTER_ID, filter_id);
                values.put(KEY_VALUE, s);
                db.replaceOrThrow(CONTENT_STRINGS_TABLE, null, values);
            }

            db.setTransactionSuccessful();
        }
        finally
        {
            db.endTransaction();
            db.close();
        }
    }

    public void deleteFilter(String name)
    {
        SQLiteDatabase db = getWritableDatabase();
        db.delete(
            FILTERS_TABLE,
            KEY_NAME + "=?",
            new String[] { name });
        db.close();
    }

    public Message getMessage(long id)
    {
        Message message = null;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.query(
            MESSAGES_TABLE,
            new String[] { KEY_ID, KEY_ADDRESS, KEY_RECEIVED_AT, KEY_VALUE },
            KEY_ID + "=?",
            new String[] { String.valueOf(id) },
            null, null, null);
        while (cursor.moveToNext())
        {
            if (message != null)
            {
                // There must be only one message with this identifier.
                throw new AssertionError();
            }
            message = getMessageFromCursor(cursor);
        }
        cursor.close();
        db.close();
        if (message == null)
            throw new AssertionError();
        return message;
    }

    public List<Message> getMessages()
    {
        List<Message> messages = new ArrayList<Message>();
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.query(
            MESSAGES_TABLE,
            new String[] { KEY_ID, KEY_ADDRESS, KEY_RECEIVED_AT, KEY_VALUE },
            null, null, null, null,
            MESSAGES_ORDER_BY);
        while (cursor.moveToNext())
            messages.add(getMessageFromCursor(cursor));
        cursor.close();
        db.close();
        return messages;
    }

    public long saveMessage(String address, long receivedAt, String message)
    {
        if (address.length() == 0)
            throw new AssertionError();
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(KEY_ADDRESS, address);
        values.put(KEY_RECEIVED_AT, receivedAt);
        values.put(KEY_VALUE, message);
        long messageID = db.replaceOrThrow(MESSAGES_TABLE, null, values);
        db.close();
        if (messageID < 0)
            throw new AssertionError();
        Intent intent = new Intent(ACTION_NEW_MESSAGE);
        mContext.sendBroadcast(intent);
        
        if(this.saveMessages())
        {
        	showMessageNotification(messageID, address, message);
        }
        return messageID;
    }

    public void deleteMessage(long id)
    {
        SQLiteDatabase db = getWritableDatabase();
        db.delete(
            MESSAGES_TABLE,
            KEY_ID + "=?",
            new String[] { String.valueOf(id) });
        db.close();

        // If there was any notification for this message, it is no longer
        // relevant. Hence, remove it.
        Notifier.cancel(mContext, String.valueOf(id), Notifier.NEW_MESSAGE);
    }

    private Message getMessageFromCursor(Cursor cursor)
    {
        long id = cursor.getLong(0);
        String address = cursor.getString(1);
        long receivedAt = cursor.getLong(2);
        String value = cursor.getString(3);
        return new Message(id, address, receivedAt, value);
    }

    private void showMessageNotification(long messageID, String address, String message)
    {
        Notification notification = Notifier.build(
                R.drawable.ic_launcher,
                "Blocked - " + address);
        Intent messageViewerIntent = new Intent(mContext, MessageViewer.class);
        messageViewerIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        messageViewerIntent.putExtra(MessageViewer.MESSAGE_ID_EXTRA, messageID);
        
        //PendingIntent pendingIntent = PendingIntent.getActivity(mContext, MessageViewer.REQUEST_CODE_MUTATED, messageViewerIntent, PendingIntent.FLAG_ONE_SHOT);
        
        
        long mod = (long)(1L<<28);
        int uniqueRequestId = (int)( messageID %  mod );
        
        PendingIntent pendingIntent = PendingIntent.getActivity(mContext, uniqueRequestId, messageViewerIntent, PendingIntent.FLAG_ONE_SHOT);
        
        notification.setLatestEventInfo(
            mContext,
            "Blocked - " + address,
            message,
            pendingIntent);
        Notifier.notify(mContext, String.valueOf(messageID), Notifier.NEW_MESSAGE, notification);
    }
}
